src/library.js

```diff
@@ -3272,8 +3272,10 @@ addToLibrary({
 #if !EXIT_RUNTIME && ASSERTIONS
     warnOnce('emscripten_force_exit cannot actually shut down the runtime, as the build does not have EXIT_RUNTIME set');
 #endif
-#if !MINIMAL_RUNTIME
+#if isSymbolNeeded('$noExitRuntime')
```

Comment:
Continuing the conversation from before, but at the place I am confused about:

Doesn't this need to know if a dependency brings in `noExitRuntime`? Say it isn't exported by the user or added in a way that `isSymbolNeeded` is aware of, but we include some function `foo` that depends on `noExitRuntime`. Then wouldn't we need to be aware of that here so that we enter the line below us?

Reply:
Yes, that would not be supported by this mechanism.. because can't know if `noExitRuntime` is pulled in via that method at the time when process the library code.   

We don't know the dependency info until later.

So code that puts `noExitRuntime` in `__deps` would not trigger the usage here.  That is a limitation of this system.

We could work around it by error'ing out in that case (i.e. making in impossible to `__deps` on `noExitRuntime`).  That wouldn't be a breaking change since `noExitRuntime` is not currently a library symbol.

Reply:
Alternatively I could just push for complete removal of `noExitRuntime` rather than this intermediate step.. but this step seems safer/smaller and that works is already done.

Reply:
I am just a little confused as to what this enables, then. Making it a library symbol enables us to depend on it as a dependency, but if that does not work, then we don't seem to benefit. Are there important use cases that do *not* use the dependency system but do use noExitRuntime? (If not, I'd prefer to just remove it directly, but if there are, sgtm.)

Reply:
This change enabled folks to continue to use `noExitRuntime` via `DEFAULT_LIBRARY_FUNCS_TO_INCLUDE`, `EXPORTED_RUNTIME_METHODS`  or `INCOMING_MODULE_JS_API` command line flags.

Since its part of the default `INCOMING_MODULE_JS_API` we actually still include it by default.

So there are several ways to include it.. just don't via the `__deps` mechanism.

There are two followups that I can see to this:

1. Reduce or empty to the default `INCOMING_MODULE_JS_API`.   This would be good code size win.
2. Completely remove `noExitRuntime`.  This would be a little more serious of breaking change since there would  be no workaround from the command line.

Reply:
Fair enough, I see how this could be used.

Do you think it's worth the extra complexity in jsifier.js for example? Moving this to a library function that also can't be used as a dep is a new kind of state for a thing to be in.

Instead, rather than using the JS library, can we just ifdef on including noExitRuntime using the same non-dep method as in this PR? That is,
```js
#ifdef needNoExitRuntime()
var noExitRuntime;
#endif
```

Reply:
Hmm.. not a bad idea.  I'll see how that looks.

Reply:
I took a stab at it but I didn't like it as much for a couple of reasons:

1. It adds complexity to `preamble.js` .. which is a file I'm trying to clean up as much as possible.
2. It means folks can't use `DEFAULT_LIBRARY_FUNCS_TO_INCLUDE` to include it.. which means they only have `EXPORTED_RUNTIME_METHODS` which has the side effect of also exporting the symbol.

So I think this somewhat hacky solution is the best approach for now.

Reply:
Why can't they use `DEFAULT_LIBRARY_FUNCS_TO_INCLUDE`? The check for whether to add `var noExitRuntime` can check that.

Reply:
Because `DEFAULT_LIBRARY_FUNCS_TO_INCLUDE` can only include symbols that the are know about by the library system.

```
$ ./emcc ~/test/hello.c -sDEFAULT_LIBRARY_FUNCS_TO_INCLUDE=xxx
error: undefined symbol: xxx (referenced by root reference (e.g. compiled C/C++ code))
```

Reply:
Ah, now I see, thanks.

Well, I don't feel strongly, so this approach sgtm. Reading again now.